### PR TITLE
Add cache read rule in auto schedule

### PR DESCRIPTION
**Pull request by BiynXu**
This pr add the AddCacheRead rule in auto schedule.
The conditions for applying this rule are: 
    1. MultiLevelTiling is required for the block.
    2. There is no cross thread reduction in the calculation of the block.
When applying this rule, we first analyze which read buffers can be cached, then call CacheRead schedule primitives one by one, and place them in the first reduce axis.
The following problems remain: 
    1. The accuracy test needs to be added. At present, the test after Multilevelling can be passed on the CPU, but the memory is out of bounds after calling AddCacheRead; The original matrix multiplication cannot be passed on the GPU, we need to debug the test code.
    2. The attribute is_reduce_axis of Var in loop node is invalid, Therefore, the way to obtain the target loop temporarily is to judge whether the Var is reduce by the name prefix.

### CODE REVIEW

File: cinn/auto_schedule/search_space/auto_gen_rule/add_cache_read.h [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+  // Returns the name of the rule, used for debug.
+  std::string GetRuleName() const override { return "AddCacheRead"; }
+
+  // Returns true if the schedule block expr is applicable by AddCacheRead
+  bool MeetCondition(const ir::Expr& block_expr) const;
+
+  // Get the out most reduce loop to set cache block in.
+  ir::Expr GetTargetLoop(const ir::Expr& block_expr) const;
+
```
Comment (CtfGo) [inline cinn/auto_schedule/search_space/auto_gen_rule/add_cache_read.h]: 不需要暴露的接口声明为private

File: cinn/auto_schedule/search_space/auto_gen_rule/add_cache_read.h [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+ private:
+  std::vector<Expr> applicable_schedule_blocks_;
```
Comment (CtfGo) [inline cinn/auto_schedule/search_space/auto_gen_rule/add_cache_read.h]:
```suggestion
  std::vector<ir::Expr> applicable_schedule_blocks_;
```
Expr与本文件中别处写法保持一致

File: cinn/auto_schedule/search_space/auto_gen_rule/add_cache_read.cc [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+#include "cinn/ir/ir_base.h"
+#include "cinn/ir/ir_printer.h"
+#include "cinn/ir/ir_schedule.h"
+#include "cinn/ir/tensor.h"
+#include "cinn/optim/ir_copy.h"
+
+namespace cinn {
+namespace auto_schedule {
+
+RuleApplyType AddCacheRead::Init(ir::IRSchedule* ir_schedule) {
```
Comment (CtfGo) [inline cinn/auto_schedule/search_space/auto_gen_rule/add_cache_read.cc:41]: 关键实现代码添加注释

File: cinn/auto_schedule/search_space/auto_gen_rule/add_cache_read.cc [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+  for (int read_buffer_index : read_buffer_indexes) {
+    ir::Expr cache_block = ir_schedule_->CacheRead(sch_block_expr, read_buffer_index, cache_memory_type_);
+    VLOG(6) << "cache block: " << cache_block;
+    sch_block_expr       = ir_schedule_->GetBlock(block_name);
+    ir::Expr target_loop = GetTargetLoop(sch_block_expr);
+    ir_schedule_->ComputeAt(cache_block, target_loop);
+  }
+}
+
+// TODO: Merge this function and the same function in MultiLevelTiling rule
```
Comment (CtfGo) [inline cinn/auto_schedule/search_space/auto_gen_rule/add_cache_read.cc]:
```suggestion
// TODO(author): detail
```

File: cinn/auto_schedule/search_space/auto_gen_rule/add_cache_read.cc [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+  }
+  for (int idx : read_buffer_indexes) {
+    VLOG(6) << "only read tensors: " << read_tensor_strs[idx];
+  }
+
+  // Do schedule
+  for (int read_buffer_index : read_buffer_indexes) {
+    ir::Expr cache_block = ir_schedule_->CacheRead(sch_block_expr, read_buffer_index, cache_memory_type_);
+    VLOG(6) << "cache block: " << cache_block;
+    sch_block_expr       = ir_schedule_->GetBlock(block_name);
```
Comment (CtfGo) [inline cinn/auto_schedule/search_space/auto_gen_rule/add_cache_read.cc:93]: 备注这里需重新getblock的原因

File: cinn/auto_schedule/search_space/auto_gen_rule/add_cache_read.cc [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+  }
+
+  std::unordered_set<std::string> write_tensor_strs;
+  ir::CollectIRNodesWithoutTensor(sch_block->body, [&](const Expr* x) {
+    if (x->As<ir::Store>()) write_tensor_strs.insert(x->As<ir::Store>()->tensor.as_tensor()->name);
+    return x->As<ir::Store>() != nullptr;
+  });
+  for (const auto& write_tensor_str : write_tensor_strs) {
+    VLOG(6) << "write tensors: " << write_tensor_str;
+  }
```
Comment (CtfGo) [inline cinn/auto_schedule/search_space/auto_gen_rule/add_cache_read.cc]:
```suggestion
if (VLOG_IS_ON(6)) {
  for (const auto& write_tensor_str : write_tensor_strs) {
    VLOG(6) << "write tensors: " << write_tensor_str;
  }
}
```
对于可能影响性能的debug日志，可以添加VLOG_IS_ON进行屏蔽

File: cinn/auto_schedule/search_space/auto_gen_rule/add_cache_read_test.cc [hunk 7]
```diff
@@ -0,0 +1,351 @@
...
+#include <iostream>
+#include <vector>
+
+#include "cinn/auto_schedule/search_space/auto_gen_rule/auto_gen_rule.h"
+#include "cinn/auto_schedule/search_space/auto_gen_rule/multi_level_tiling.h"
+#include "cinn/backends/codegen_c.h"
+#include "cinn/backends/codegen_cuda_dev.h"
+#include "cinn/backends/compiler.h"
+#include "cinn/backends/nvrtc_util.h"
+#include "cinn/cinn.h"
```
Comment (CtfGo) [inline cinn/auto_schedule/search_space/auto_gen_rule/add_cache_read_test.cc:28]: 包含了cinn.h之后下面很多冗余的头文件可以删去

File: cinn/ir/ir.h [hunk 8]
```diff
@@ -709,6 +709,8 @@ struct ForBase {
   inline bool is_binded() const {
     return tell_for_type_flag(ForType::GPUBlock) || tell_for_type_flag(ForType::GPUThread);
   }
+  inline bool is_GPUBlockBinded() const { return tell_for_type_flag(ForType::GPUBlock); }
```
Comment (CtfGo) [inline cinn/ir/ir.h]: 要么用驼峰规范，要么与别处一致全小写；另外，这两个接口是否可能被多处使用，如果仅是该pr需要，可以加在add_cache_read.cc中作为本地函数

File: cinn/auto_schedule/search_space/auto_gen_rule/add_cache_read_test.cc [hunk 9]
```diff
@@ -0,0 +1,351 @@
...
+// TODO: Debug and check result on cuda
+void check_matmul_result_cuda(int M, int N, int K, void (*func_ptr)(void**, int32_t)) {
+  // prepare data
+  auto* A_host = common::BufferBuilder(Float(32), {M, N}).set_random().Build();
+  CHECK(A_host);
+  auto* B_host = common::BufferBuilder(Float(32), {M, N}).set_random().Build();
+  CHECK(B_host);
+  auto* C_host = common::BufferBuilder(Float(32), {M, N}).set_zero().Build();
+  CHECK(C_host);
+  auto* C_res_host = common::BufferBuilder(Float(32), {M, N}).set_zero().Build();
```
Comment (CtfGo) [inline cinn/auto_schedule/search_space/auto_gen_rule/add_cache_read_test.cc]: 这个C_res_host什么用处？

File: cinn/auto_schedule/search_space/auto_gen_rule/add_cache_read_test.cc [hunk 10]
```diff
@@ -0,0 +1,351 @@
...
+  EXPECT_EQ(add_cache_read.Init(&ir_schedule), RuleApplyType::kApplyAndSkipAllRules);
+  EXPECT_EQ(add_cache_read.NumberApplicable(), 1);
+
+  add_cache_read.ApplyRandomly();
+  exprs = ir_schedule.GetModule().GetExprs();
+  EXPECT_EQ(exprs.size(), 1UL);
+  VLOG(6) << "Expr after AddCacheRead: " << exprs[0];
+
+  auto temp_buffers = lang::GetTempBuffers({A, B, C}, stages, exprs[0]);
+  auto func         = ir::_LoweredFunc_::Make(funcs[0]->name, funcs[0]->args, funcs[0]->body, temp_buffers);
```
Comment (CtfGo) [inline cinn/auto_schedule/search_space/auto_gen_rule/add_cache_read_test.cc]:
```suggestion
  auto func         = ir::_LoweredFunc_::Make(funcs[0]->name, funcs[0]->args, exprs[0], temp_buffers);
```
此处应为exprs[0]？

Comment (BiynXu) [reply cinn/auto_schedule/search_space/auto_gen_rule/add_cache_read_test.cc]: 已修改

File: cinn/auto_schedule/search_space/auto_gen_rule/add_cache_read_test.cc (hunk 9 above)
Comment (BiynXu) [reply cinn/auto_schedule/search_space/auto_gen_rule/add_cache_read_test.cc]: 已删除

File: cinn/ir/ir.h (hunk 8 above)
Comment (BiynXu) [reply cinn/ir/ir.h]: 已修改为一致小写

File: cinn/auto_schedule/search_space/auto_gen_rule/add_cache_read_test.cc (hunk 7 above)
Comment (BiynXu) [reply cinn/auto_schedule/search_space/auto_gen_rule/add_cache_read_test.cc:28]: done

File: cinn/auto_schedule/search_space/auto_gen_rule/add_cache_read.cc (hunk 6 above)
Comment (BiynXu) [reply cinn/auto_schedule/search_space/auto_gen_rule/add_cache_read.cc]: done

File: cinn/auto_schedule/search_space/auto_gen_rule/add_cache_read.cc (hunk 5 above)
Comment (BiynXu) [reply cinn/auto_schedule/search_space/auto_gen_rule/add_cache_read.cc:93]: 由于CacheRead中生成了新block替换，原block失效，因此此处重新按名字Get。已在代码中注释。

File: cinn/auto_schedule/search_space/auto_gen_rule/add_cache_read.cc (hunk 4 above)
Comment (BiynXu) [reply cinn/auto_schedule/search_space/auto_gen_rule/add_cache_read.cc]: done

File: cinn/auto_schedule/search_space/auto_gen_rule/add_cache_read.h (hunk 2 above)
Comment (BiynXu) [reply cinn/auto_schedule/search_space/auto_gen_rule/add_cache_read.h]: done

File: cinn/auto_schedule/search_space/auto_gen_rule/add_cache_read.h (hunk 1 above)
Comment (BiynXu) [reply cinn/auto_schedule/search_space/auto_gen_rule/add_cache_read.h]: done

File: cinn/auto_schedule/search_space/auto_gen_rule/add_cache_write.cc [hunk 11]
```diff
@@ -77,57 +77,11 @@ void AddCacheWrite::Apply(int index) {
...
-}
-
 bool AddCacheWrite::MeetCondition(const ir::Expr& block_expr) const {
   const ir::ScheduleBlockRealize* sch_block_realize = block_expr.As<ir::ScheduleBlockRealize>();
   const ir::ScheduleBlock* sch_block                = sch_block_realize->schedule_block.As<ir::ScheduleBlock>();
 
-  if (sch_block->read_buffers.empty() || sch_block->write_buffers.size() != 1) {
-    return false;
-  }
-
```
Comment (CtfGo) [inline cinn/auto_schedule/search_space/auto_gen_rule/add_cache_write.cc:127]: 这个条件不需要了吗？

Comment (BiynXu) [reply cinn/auto_schedule/search_space/auto_gen_rule/add_cache_write.cc:127]: 在NeedsMultiLevelTiling函数中有这个条件了

Comment (CtfGo) [reply cinn/auto_schedule/search_space/auto_gen_rule/add_cache_write.cc:127]: ok

**Review by CtfGo — APPROVED** [review]
LGTM

**Pull request merged**